src/Table.cpp

```diff
@@ -359,191 +433,440 @@ namespace galsim {
+                                     _dfdx[(i-1)*_ny+j]*dxgrid, _dfdx[i*_ny+j]*dxgrid);
+            double der0 = oneDSpline(dx, _dfdy[(i-1)*_ny+j-1], _dfdy[i*_ny+j-1],
+                                     _d2fdxdy[(i-1)*_ny+j-1]*dxgrid, _d2fdxdy[i*_ny+j-1]*dxgrid);
+            double der1 = oneDSpline(dx, _dfdy[(i-1)*_ny+j], _dfdy[i*_ny+j],
+                                     _d2fdxdy[(i-1)*_ny+j]*dxgrid, _d2fdxdy[i*_ny+j]*dxgrid);
+
+            return oneDSpline(dy, val0, val1, der0*dygrid, der1*dygrid);
+        }
+
+        void grad(double x, double y, int i, int j, double& dfdx, double& dfdy) const {
+            double dxgrid = _xargs[i] - _xargs[i-1];
+            double dygrid = _yargs[j] - _yargs[j-1];
+            double dx = (x - _xargs[i-1])/dxgrid;
+            double dy = (y - _yargs[j-1])/dygrid;
+
+            // xgradient;
+            double val0 = oneDGrad(dx, _vals[(i-1)*_ny+j-1], _vals[i*_ny+j-1],
+                                   _dfdx[(i-1)*_ny+j-1]*dxgrid, _dfdx[i*_ny+j-1]*dxgrid);
+            double val1 = oneDGrad(dx, _vals[(i-1)*_ny+j], _vals[i*_ny+j],
+                                   _dfdx[(i-1)*_ny+j]*dxgrid, _dfdx[i*_ny+j]*dxgrid);
+            double der0 = oneDGrad(dx, _dfdy[(i-1)*_ny+j-1], _dfdy[i*_ny+j-1],
+                                   _d2fdxdy[(i-1)*_ny+j-1]*dxgrid, _d2fdxdy[i*_ny+j-1]*dxgrid);
+            double der1 = oneDGrad(dx, _dfdy[(i-1)*_ny+j], _dfdy[i*_ny+j],
+                                   _d2fdxdy[(i-1)*_ny+j]*dxgrid, _d2fdxdy[i*_ny+j]*dxgrid);
+            dfdx = oneDSpline(dy, val0, val1, der0*dygrid, der1*dygrid)/dxgrid;
+
+            // ygradient
+            val0 = oneDGrad(dy, _vals[(i-1)*_ny+j-1], _vals[(i-1)*_ny+j],
+                            _dfdy[(i-1)*_ny+j-1]*dygrid, _dfdy[(i-1)*_ny+j]*dygrid);
+            val1 = oneDGrad(dy, _vals[i*_ny+j-1], _vals[i*_ny+j],
+                            _dfdy[i*_ny+j-1]*dygrid, _dfdy[i*_ny+j]*dygrid);
+            der0 = oneDGrad(dy, _dfdx[(i-1)*_ny+j-1], _dfdx[(i-1)*_ny+j],
+                            _d2fdxdy[(i-1)*_ny+j-1]*dygrid, _d2fdxdy[(i-1)*_ny+j]*dygrid);
+            der1 = oneDGrad(dy, _dfdx[i*_ny+j-1], _dfdx[i*_ny+j],
+                            _d2fdxdy[i*_ny+j-1]*dygrid, _d2fdxdy[i*_ny+j]*dygrid);
+            dfdy = oneDSpline(dx, val0, val1, der0*dxgrid, der1*dxgrid)/dygrid;
+        }
+    private:
+
+        double oneDSpline(double x, double val0, double val1, double der0, double der1) const {
+            // assuming that x is between 0 and 1, val0 and val1 are the values at
+            // 0 and 1, and der0 and der1 are the derivatives at 0 and 1.
+
+            // I'm assuming the compiler will reduce this all down...
+            double a = 2*(val0-val1) + der0 + der1;
+            double b = 3*(val1-val0) - 2*der0 - der1;
+            double c = der0;
+            double d = val0;
+
+            return d + x*(c + x*(b + x*a));
+        }
```

Comment:
Also, it would probably be worth an inline comment about the math you are doing here. This calculation isn't "true by inspection". (At least for me.)  This might help make clear the connection to how this relates to the 1D spline calculation, which looks very different.

Reply:
I think this is the difference between a "natural" spline (what `setupSpline` does) and a "Catmull-Rom" spline.  The natural spline is smoother, since it enforces the continuous first derivatives.  To get that smoothness though, you have to solve ~N simultaneous (though  sparse) equations.  This isn't so bad in 1D - it's inverting a tri-diagonal matrix.  In 2+D though, you lose the tridiagonality since points in more than one dimension aren't well-ordered.  It's still sparse though, so I'm sure it's still possible, but I didn't try it.  (...and all the bicubic spline algorithms I found online were either the Catmull-Rom type, or the cubic convolution, which makes a third variant).

So it is a different design and will yield different results than what we do in 1D.  I think that's okay though; I don't think I'd advocate changing either one.

FWIW, one potential advantage of the Catmull-Rom spline is that it's local.  Changing the value of a point only changes the interpolation near that point, which isn't true for the natural spline.

I did go and make some additional comments to elucidate the maths here.